### PR TITLE
Registry function returns ResolvedValue

### PR DESCRIPTION
**Pull request by mvilks**

### CODE REVIEW

File: template/example_test.go [hunk 1]
```diff
@@ -50,44 +50,51 @@ func ExampleTemplate_complexMessage() {
...
 
 		color, ok := value.(string)
 		if !ok {
-			return nil, fmt.Errorf("input is not a string: %w", mf2.ErrBadOperand)
+			return errorf("input is not a string: %w", mf2.ErrBadOperand)
 		}
 
 		if options == nil {
-			return color, nil
+			return template.NewResolvedValue(color), nil
```
Comment (janishorsts) [inline template/example_test.go:68]: no format?

Comment (mvilks) [reply template/example_test.go:68]: It uses defaultFormat while it still can (until we start throwing errors).

File: template/example_test.go [hunk 2]
```diff
@@ -50,44 +50,51 @@ func ExampleTemplate_complexMessage() {
 .input { $color :color style=RGB}
 {{John is { $age } years old and his favorite color is { $color }.}}`
 
-	color := func(value any, options template.Options, locale language.Tag) (any, error) {
+	color := func(value any, options template.Options, locale language.Tag) (*template.ResolvedValue, error) {
+		errorf := func(format string, args ...any) (*template.ResolvedValue, error) {
+			return template.NewResolvedValue(""), fmt.Errorf("exec color function: "+format, args...)
```
Comment (janishorsts) [inline template/example_test.go:55]: Does this still work? Do registry func consumers care about returned value on error?

```suggestion
			return nil, fmt.Errorf("exec color function: "+format, args...)
```

File: template/registry_datetime.go [hunk 3]
```diff
@@ -179,19 +179,23 @@ func parseDatetimeOptions(options Options) (*datetimeOptions, error) {
...
+
 	value, err := parseDatetimeOperand(operand)
 	if err != nil {
-		return "", fmt.Errorf("exec datetime func: %w", err)
+		return errorf("%w", err)
 	}
 
 	if len(options) == 0 {
-		return fmt.Sprint(operand), nil // TODO(mvilks): should return format()
+		return NewResolvedValue(operand), nil
```
[the remark this answers is not in the document]
Comment (janishorsts) [reply template/registry_datetime.go:193]: Is this _if_ statement even needed?

File: template/registry_string.go [hunk 4]
```diff
@@ -8,32 +8,36 @@ import (
...
-	errorf := func(format string, args ...any) (any, error) {
-		return nil, fmt.Errorf("exec string function: "+format, args...)
+func stringFunc(operand any, options Options, locale language.Tag) (*ResolvedValue, error) {
+	errorf := func(format string, args ...any) (*ResolvedValue, error) {
+		return NewResolvedValue(""), fmt.Errorf("exec string function: "+format, args...)
 	}
 
 	if operand == nil {
-		return "", nil
+		return NewResolvedValue(""), nil
```
Comment (janishorsts) [inline template/registry_string.go:17]: missing format func

File: template/example_test.go (hunk 2 above)
Comment (mvilks) [reply template/example_test.go]: Good point, replaced.

File: template/example_test.go (hunk 1 above)
Comment (janishorsts) [reply template/example_test.go:68]: We know we'll add erroring later and ResolvedValue MUST have at least one - formatting or selection. Let's make sure we return a valid (error free) ResolvedValue.

File: template/registry_datetime.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mvilks) [reply template/registry_datetime.go:193]: Per documentation, without options the `:datetime` is the same as `:datetime datestyle=medium timestyle=short`. Fixed accordingly.

File: template/example_test.go (hunk 1 above)
Comment (mvilks) [reply template/example_test.go]: Added format function

File: template/example_test.go [hunk 5]
```diff
@@ -50,44 +50,51 @@ func ExampleTemplate_complexMessage() {
...
 
 		color, ok := value.(string)
 		if !ok {
-			return nil, fmt.Errorf("input is not a string: %w", mf2.ErrBadOperand)
+			return errorf("input is not a string: %w", mf2.ErrBadOperand)
 		}
 
 		if options == nil {
-			return color, nil
+			return template.NewResolvedValue(color, template.WithFormat(func() string { return color })), nil
```
Comment (janishorsts) [inline template/example_test.go:68]: move it to format func instead?

File: template/example_test.go [hunk 6]
```diff
@@ -50,44 +50,51 @@ func ExampleTemplate_complexMessage() {
...
 		}
 
-		style, err := options.GetString("style", "RGB")
-		if err != nil {
-			return nil, fmt.Errorf("get style: %w", err)
-		}
+		format := func() string {
+			style, err := options.GetString("style", "RGB")
```
Comment (janishorsts) [inline template/example_test.go:72]:
```suggestion
		format := func() string {
			if len(options) == nil {
				return color
			}

			style, err := options.GetString("style", "RGB")
```

File: template/registry_date_test.go [hunk 7]
```diff
@@ -54,6 +51,7 @@ func Test_Date(t *testing.T) {
 				t.Error(err)
 			}
 
+			got := v.format()
```
[the remark this answers is not in the document]
Comment (mvilks) [reply template/registry_date_test.go:54]: It should `t.Error()` if dateFunc returns err.

File: template/template.go [hunk 8]
```diff
@@ -37,6 +37,11 @@ type ResolvedValue struct {
 	err       error
 }
 
+// Value returns raw value.
+func (r *ResolvedValue) Value() any {
```
Comment (janishorsts) [inline template/template.go:41]: unused?

File: template/template_test.go [hunk 9]
```diff
@@ -44,7 +44,7 @@ func Test_ExecuteSimpleMessage(t *testing.T) {
 			name: "function without operand",
 			text: "Hello, { :randName }",
 			funcs: Registry{
-				"randName": func(any, Options, language.Tag) (any, error) { return "John", nil },
+				"randName": func(any, Options, language.Tag) (*ResolvedValue, error) { return NewResolvedValue("John"), nil },
```
Comment (janishorsts) [inline template/template_test.go:47]: Is it gonna break when we add error on missing format function?

File: template/template.go (hunk 8 above)
Comment (mvilks) [reply template/template.go:41]: It's for external devs to get to the raw value.

Comment (janishorsts) [reply template/template.go:41]: internally, we access ResolvedValue.value

Comment (janishorsts) [reply template/template.go:41]: Does the current implementation require it? This PR. :)

File: template/registry_date_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (janishorsts) [reply template/registry_date_test.go:56]: t.Error() continues execution.

**Review by janishorsts — APPROVED** [review]

**Pull request merged**